### ISSUE TITLE
Stamina drain duration

### ISSUE THREAD

**Issue by Pegwald** [aura-project/aura#294]
I noticed if you use a skill like playing instrument or refining, stamina will be consumed rapidly in an instant instead of over time. Not just life skills, but also combat. I'm actually pretty sure this goes the same way for magic.
My point is that some skills should drain the required amount immediately, and some should drain it over time for a **longer** duration.

Loading combat skills should drain the stamina instantly, like smash. I noticed that it drains stamina for a bit, then it stops. For skills like Counter, it gets a little confusing if this current drain effect is overlapping the regular drain that subtracts stamina while having the skill active. I also have a theory that you can avoid paying some the stamina cost if you load a skill like smash, then use it immediately. That's just a theory so far.
 
Also worth mentioning - According to Wiki, playing instrument should consume 0.5 SP/sec, but instead it drains 5 stamina rapidly over 1 sec. Most songs last much longer than 10 sec, so 5 stamina as a static cost (no matter the length of the song) is really cheap then, right? Not really, because right now if you try to rank "Playing Instrument" (going for perfect performances) you will pay 5 stamina per attempt, even if you cancel the skill or not. That's a lot of drain to get a few perfect performances.

I'm not sure if you're also aware of this already, but anyway, I felt it was worth making an issue out of this just in case.

**Comment by Pegwald** [comment]
I just noticed I might be wrong about the playing instrument drain, because I know that playing **without** a scroll consumes stamina, playing **with** a scroll doesn't. I am not sure if the 0,5/sec count towards playing with or without a scroll, so now I feel a bit awkward.

**Comment by exectails** [comment]
I see one potential issue with the skill handling, which is that `staminaActive` is reduced in one second on SkillUse. This was changed when Lightning Rod was added, because it was supposed to drain `manaActive` in one second.

If this is incorrect, and the `Active` regen should actually be active until Complete is sent, which seems to be the case for skills like Enchant and Playing Instrument, then we need something to make sure a regen ticks at least once, otherwise a regen might get removed before it reduces the stat.

At the same time, lag could cause the regen to run too long though... that's a problem. Hm, this actually sparks an interesting question, how long do skills run? Is there a way for us to get the time the skill will be "active"? The client must know somehow. If we know the duration of Use, until Complete is sent, we can accurately set the regen.

>playing instrument should consume 0.5 SP/sec, but instead it drains 5 stamina rapidly over 1 sec. 

Going by the data it seems like it should use 5 Stamina *initially*, which is the amount you need to use the skill at all, and then it uses .5 for the duration of the play. That second part doesn't work atm because those .5 are drained in 1s, you don't see it.

```js
staminaCost: 5, staminaPrepare: -5, staminaWait: 0, staminaActive: -0.5
```

(xCost is the value we currently check to see if the skill can be used at all, it's usually equal to xPrepare. The negative values is what is actually taken afterwards.)

>For skills like Counter, it gets a little confusing if this current drain effect is overlapping the regular drain that subtracts stamina while having the skill active.

What do you mean by this? Looks like Counter should drain 2 Stamina on Prepare and then 1 per second while you're waiting.

```js
staminaCost: 2, staminaPrepare: -2, staminaWait: -1, staminaActive: 0
```

Since Counter doesn't have a load time anymore those two would overlap each other, but is that a problem?

**Comment by Pegwald** [comment]
> That second part doesn't work atm because those .5 are drained in 1s, you don't see it.

_Something_ is at least draining the stamina bar while playing instrument, so now I am a bit confused about what it might be. The wiki doesn't mention any other stamina costs than the 0.5/per sec, which does seem to work, its just not making any sense to me why it also costs 5 stamina to load the skill :/
Is there something I have missed?


> Since Counter doesn't have a load time anymore those two would overlap each other, but is that a problem?

It's not _really_ a problem, it's just different from what you see in the official version (or from what I can remember). I'm sorry if i'm wrong about this, but I can seem to remember the cost for activating the skill being subtracted immediately, then adding stamina drain per sec.
The stamina being drained the way it does now is really the reason why I made this issue. Sometimes it should just be a longer drain duration and sometimes just be an instant subtract.

I don't know what changed in NA since g18, so after all there revamps I might be wrong about this entire issue. I will log on today for a moment of clarity.

**Comment by exectails** [comment]
Could you clarify what you mean by
>instant subtract
?

As far as I knew there should *always* be a drain, aka it looks animated instead of just jumping from say 100 to 90 in an instant, with the only difference being *how long* the regen runs. As regens tick (activate) once every second, the only way to remove life/mana/stamina with an animated bar would be to set up a regen with a duration of 1s, so it ticks once, at which point the client animates the removal of the amount, or without a duration to keep it running until you stop it.

However, I just tried Counterattack on NA and noticed the initial costs actually *are* reduced without a regen... the value is simply set =< That's incorrect on Aura, you're right.

I also just realized Playing Instrument doesn't use the Ready and Use handlers, it goes straight to Use, and the regen is set up in the skill's Prepare handler. That explains why you saw a constant drain on Aura. Which is correct I guess, even though initially it sounded like this *didn't* work.

>playing instrument should consume 0.5 SP/sec, but *instead* it drains 5 stamina rapidly over 1 sec

So the problem is the 5 Stamina, which aren't removed on officials... I think this could mean two things, 1) we're setting up the Prepare cost regen in the wrong location, or 2) the costs are omitted if the skill goes straight to Use.

Further you're kinda correct about the scroll thing, the 0.5 drain only applies to playing *without* a scroll. *With* a scroll you apparently don't lose *any* Stamina.

Proposed fixes for all bugs:
- Instant stat reduction for Prepare.
- No instant stat reduction for Prepare for straight to Use skills?
- Only drain Stamina for Playing Instrument if playing without scroll.

**Comment by exectails** [comment]
Correction: You seem to always instantly lose .5 on activation of Playing Instrument, but the *drain* is no-scroll exclusive.

**Comment by Pegwald** [comment]
> instant subtract

What I meant was in some cases there are just a single tick were all the required stamina is drained instantly - the moment you active a skill. I just thought there was no animation of the subtraction of stamina/mana when the costs were too low, but high costs like meteor, ice spear or other heavy consuming mana/stamina skills has "bar-animations".
This might be off track, but the draining from ice spear is very fun to watch from first to last charge...

> problem is the 5 Stamina, which aren't removed on officials

Really? The wiki doesn't mention the cost at all, but I guess the wiki doesn't really state everything correctly sometimes... x-x'

> Correction: You seem to always instantly lose .5 on activation of Playing Instrument, but the drain is no-scroll exclusive.

That sort of corrects my theory. I knew there were different costs if you were playing with or without a scroll.

**Comment by exectails** [comment]
> I just thought there was no animation of the subtraction of stamina/mana when the costs were too low, but high costs like meteor, ice spear or other heavy consuming mana/stamina skills has "bar-animations".

The advanced magic skills *are* animated? So only "low" amounts are removed instantly and others use regens?

**Comment by exectails** [comment]
Note: What I mean by removed instantly is for example 50 -> 45 in an *instant*, *no* animation. Which is admittedly hard to see with values like .5~2 without logging the packets.

**Comment by Pegwald** [comment]
Now we're on the same track ^^

**Comment by exectails** [comment]
I've realized it's not that *some* skills *just* instantly reduce the stats, but that they (probably) do that if the cast time is 0. The Prepare cost is supposed to be reduced over the course of loading the skill, but if that time is 0, using a regen doesn't make sense.

While researching this I realized we currently have a small bug in our Prepare regens. We run them for one second, removing xPrepare. Since our regens tick once every second, the cost is only removed once, but the cost specified in the db is not the full amount, but the cost per second, relative to the new load time.

For example, Firebolt has manaPrepare 0.83 and a newLoadTime of 1800ms. The manaCost are 1.5. 0.83 * 1.8 is 1.494, or 1.5, which is the total cost.

This means we can't simply start a regen for the duration of the cast with the Prepare cost, we have to adjust the cost for our per second tick. Alternatively we could change our regen tick to 100ms. We would still have to adjust the value, but the math would be slightly easier.
